clang/lib/Sema/SemaAPINotes.cpp

```diff
@@ -814,11 +814,42 @@ void Sema::ProcessAPINotes(Decl *D) {
 
   // Globals.
   if (D->getDeclContext()->isFileContext() ||
+      D->getDeclContext()->isNamespace() ||
       D->getDeclContext()->isExternCContext()) {
+    std::optional<api_notes::Context> APINotesContext;
+    if (auto NamespaceContext = dyn_cast<NamespaceDecl>(D->getDeclContext())) {
+      for (auto Reader :
+           APINotes.findAPINotes(NamespaceContext->getLocation())) {
+        // Retrieve the context ID for the parent namespace of the decl.
+        std::stack<NamespaceDecl *> NamespaceStack;
+        {
+          auto CurrentNamespace = NamespaceContext;
+          while (CurrentNamespace) {
+            NamespaceStack.push(CurrentNamespace);
+            CurrentNamespace =
+                dyn_cast<NamespaceDecl>(CurrentNamespace->getParent());
```

Comment:
This seems like as good a place as any to bring up the question of C++ inline namespaces. Should they be represented in the YAML at all, or should we ignore them? For example, if you want to add an API note for `std::vector`, with libc++ that's something like `std::__1::vector`, where the `__1` is an inline namespace. Should the YAML mention the `__1`, or should it ignore inline namespaces entirely?

Reply:
This is an interesting question. In the case of `std::__1` I think we definitely don't want to spell out `__1` in YAML, since the name of the nested inline namespace is different in libc++ vs libstdc++, and it would be ideal to have the same YAML file for various C++ stdlibs.

I think we should not spell out inline namespaces in YAML for now. If there is a use case for spelling them out in YAML, we can always add support for that later. For instance, for C++ type `ns1::inline_ns2::inline_ns3::MyType` we could allow referring to it as either `ns1::MyType` or `ns1::inline_ns2::inline_ns3::MyType`, that would be fairly simple to implement.

Reply:
What about disambiguation?  What happens when I want to modify `__swift::__runtime::llvm::APInt` but not `::llvm::APInt`?  This is a problem for some programming languages ;-)

Reply:
Right now that wouldn't be possible, but we could always add support for specifying inline namespaces in YAML in the future.

For instance, we could allow this syntax:
```
Namespaces:
- Name: inline __swift
  Namespaces:
  - Name: inline __runtime
    ...
```
or something like this:
```
Namespaces:
- Name: __swift
  Inline: true
  Namespaces:
  - Name: __runtime
    Inline: true
    ...
```

I don't think it's worth blocking this patch on this issue: no matter what we will decide on the question of internal namespaces, we would want to allow the current behavior of skipping inline namespaces in YAML. So, there wouldn't be source breakage if/when we introduce the new syntax for inline namespaces later.

Reply:
I kinda prefer the second approach btw, the attribute to indicate it is an inline namespace.